### DISCUSSION TITLE
失去焦点事件的处理(Handling of lost focus events)

### DISCUSSION THREAD

**Discussion by ZeronoFreya** [wailsapp/wails#1418] [Show and tell]
我感觉是个挺常用的功能，但没找到合适的例子，自己研究了一下。
需要的库：
```
"github.com/lxn/win"
```
确保在窗口创建完成之后设置，需要获得窗口句柄（Hwnd），建议在 `startup` 方法中使用

如果有不合适的地方，欢迎指出

--- 
## Google Translate
I feel that it is a very commonly used function, but I didn't find a suitable example, so I researched it myself.
Required Libraries:
```
"github.com/lxn/win"
```
Make sure to set it after the window is created, you need to get the window handle (Hwnd), it is recommended to use it in the `startup` method

If there is any inappropriate place, please point out

---

```go
import (
	"context"
	"syscall"

	"github.com/lxn/win"
	"github.com/wailsapp/wails/v2/pkg/runtime"
)
var childHwnd []win.HWND

func storeHwnd(hwnd win.HWND, lParam uintptr) uintptr {
	childHwnd = append(childHwnd, hwnd)
	return 1 // true to continue
}

func hwndNotInChild(hwnd win.HWND) bool {
	for _, v := range childHwnd {
		if v == hwnd {
			return false
		}
	}
	return true
}

func MakeWindowsMessageReceiver(ctx context.Context) win.WINEVENTPROC {
	return func(hWinEventHook win.HWINEVENTHOOK,
		event uint32, hwnd win.HWND,
		idObject int32, idChild int32,
		idEventThread uint32, dwmsEventTime uint32) uintptr {
		if hwndNotInChild(hwnd) {
			runtime.WindowHide(ctx)
		}
		return 0
	}
}

func StartWindowsMessageLoop(windowsMessageReceiver win.WINEVENTPROC) int {
	win.SetWinEventHook(win.EVENT_OBJECT_FOCUS, win.EVENT_OBJECT_FOCUS, 0, windowsMessageReceiver, 0, 0,
		win.WINEVENT_OUTOFCONTEXT|win.WINEVENT_SKIPOWNPROCESS)

	var msg win.MSG
	for {
		if win.GetMessage(&msg, 0, 0, 0) == 0 {
			break
		}
		win.TranslateMessage(&msg)
		win.DispatchMessage(&msg)
	}
	return int(msg.WParam)
}

func (a *App) startup(ctx context.Context) {
    hwnd := win.FindWindow(nil, syscall.StringToUTF16Ptr("Your app title, find in main.go"))
    childHwnd = nil
    win.EnumChildWindows(hwnd, syscall.NewCallback(storeHwnd), 0)
    go StartWindowsMessageLoop(MakeWindowsMessageReceiver(ctx))
}

```

**Comment by leaanthony** [comment]
Wails isn't really designed for this use case. Could you explain more why you are trying to open a second window and patch the event loops?

**Reply by ZeronoFreya** [reply]
Because this is a desktop program. Please don't consider the rules of the browser, qualified desktop programs need to interact with the system, including but not limited to dragging files to the window to get the file path (we sometimes don't like getting the file itself, such as various launcher applications) when used as a tool program , and also need to automatically hide when the focus is lost. These are very practical functions. If you can only do what the browser allows, why don't I just make a web page?

Of course, this is just my personal opinion
